### PR TITLE
TST: Use Actions for LTS

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to address the 4.0.x branch for #11050 .

#### Jobs not backported

* linkcheck
* ppc

#### Expected failures

- remote-data/numpy-dev -- Temporarily ignored at #11116
- `wheels_cp39*win32` -- `utils` failure; fix at #11089

#### Unexpected failures that are out of scope

Fixing these would be nice but should not block the PR.

- [x] ` cp36-macosx_x86_64 wheel` -- `[error]Exit code 134 returned from process: file name '.../bin/node', arguments '".../bash.js"` (Transient)
- [x] `cp38-macosx_x86_64` -- `hypothesis.errors.Flaky: Hypothesis test_datetime_timedelta_roundtrip(scale='ut1', days=67328, microseconds=1116906528) produces unreliable results`

#### Unexpected failures that need to be fixed

- [x] Coverage job -- `unrecognized arguments: --readonly`
- [x] Python 3.7 with Numpy 1.17 and full coverage -- Numpy API mismatch between matplotlib and numpy:
```
INTERNALERROR>   File "/home/runner/work/astropy/astropy/.tox/py37-test-alldeps-numpy117-cov-clocale/lib/python3.7/site-packages/pytest_mpl/plugin.py", line 125, in pytest_report_header
INTERNALERROR>     import matplotlib.ft2font
INTERNALERROR> ImportError: numpy.core.multiarray failed to import
RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd
```
- [x] `aarch64` and `s390x` -- missing lib

### CODE REVIEW

**Comment by pllim** [comment]
Need expertise from @astrofrog , @mhvk , or @olebole on the exotic archs:

```
    The headers or library files could not be found for zlib,
    a required dependency when compiling Pillow from source.
```

**Comment by olebole** [comment]
You need to install the development libraries of zlib. On Debian/Ubuntu:
```
apt-get install zlib1g-dev
```

**Comment by pllim** [comment]
@olebole , thanks, that solved `zlib`, but now it is complaining about `The headers or library files could not be found for jpeg`. Not sure why these being missing isn't a problem in `master` but somehow is a problem in `v4.0.x` branch.

**Comment by olebole** [comment]
@plim Similarly, install `libjpeg-dev`. Probably you will need to install even more -- see https://packages.debian.org/source/stable/pillow for a list of build dependencies ("adep") if the "pillow" package.Maybe you just install `python3-pil` from the system and this way avoid to install it yourself? It is in stable however version 5.4.1; but if that works, it may be the quicker way.

**Comment by pllim** [comment]
Not sure why Azure didn't kick off this time. 🤷

**Comment by Cadair** [comment]
> Not sure why Azure didn't kick off this time. shrug

you upset it.

**Comment by pllim** [comment]
Now the aarch64 and s390x jobs are failing at trying to build freetype...

```
    Building freetype in build/freetype-2.6.1
    GNU make (>= 3.80) or makepp (>= 1.19) is required to build FreeType2.
```

**Comment by olebole** [comment]
`apt-get install libfreetype6-dev`?

**Comment by pllim** [comment]
That didn't work but I think it is a symptom of a different problem. I don't think it should be trying to build these things from source.
```
error: invalid command 'bdist_wheel'
```

**Comment by pllim** [comment]
Yeah, I dunno what is going on with the `aarch64` and `s390x`. Ideally, I shouldn't have to modify what I copied over from `master` at all. Options:

* Wait for @astrofrog to tell me what to do.
* Remove these two archs from LTS CI. (Not ideal?)

**Comment by olebole** [comment]
@plim since we build it on 390x for Debian, I am wondering what the exact problem here is (I couldn't find it out by the logs). One  way would be to just install build dependencies from the Debian build as needed. They are [here](https://salsa.debian.org/debian-astro-team/astropy/-/blob/82665dc6fccc5578b5276a2a9232a3e9a9b3e1d3/debian/control#L6-39).

**Comment by mhvk** [comment]
Another option, maybe sounding strange, is to install `python3-astropy` - it will pull everything in. That's how I had it initially on `travis`, before we felt that it was perhaps a bit circular. But for LTS, the dependencies are going to remain OK and we can at least move forward! (just need to check the logs once that we are not testing Debian's version!)

**Comment by olebole** [comment]
If deb-src is in `etc/apt/sources.list`:
```
deb-src http://deb.debian.org/debian/ buster main
```
one could also do `apt build-dep astropy`, which pulls all dependencies to build astropy (the buster version 3.1.2).

**Comment by olebole** [comment]
I just made a test (on x86_64): I could build (no tests) the astropy-4.2 with the following updates:

* pyerfa (rebuild from the current sources; needed in turn the new liberfa, tests disabled)
* extension-helpers (rebuild from the current sources)
* numpy-1.19 (needed to update from pip). This would seriously prevent me from providing a backport btw.

Pytest-4 would be another, for the tests.

**Comment by pllim** [comment]
This PR is only for astropy 4.0.x . It has new errors now but I am not familiar with it.

**Comment by embray** [comment]
I'll have a look at this.

**Comment by pllim** [comment]
Thanks, @embray ! Might need a rebase, not sure. Feel free to take over this PR.

**Comment by embray** [comment]
I'll rebase.  Was there any remaining issue other than the failing tests on "exotic" architectures?

**Comment by embray** [comment]
Also @pllim your branch names are always such a joy.

**Comment by pllim** [comment]
> Was there any remaining issue

Yes, the non-exotic CI was failing too. Though this current error is different than what I remember -- `RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd`

**Comment by pllim** [comment]
> Not sure why clocale had matplotlib pinned to 2.1.x

I have no recollection of this. If @astrofrog doesn't know, I'll dig more tomorrow. Thanks for working on this!

**Comment by embray** [comment]
@pllim Well what's weird is this tox.ini did not even appear in this branch until this PR, but I don't see any record of that line ever existing in the tox.ini on the main branch.

Also why are you on GH on Sunday night? XD

**Comment by embray** [comment]
I'm working on the aarch64 and s390x builds.  I have a few fixes I'm testing out to speed up their builds.  I also fixed the Python 3.7+numpy 1.17 build but I'll push the fix tomorrow once I (hopefully) have the other fixes working as well.  Gonna go to bed and check in the morning since it's still quite slow to run the tests on those =)

**Comment by embray** [comment]
aarch64 and s390x are working now.  @pllim you mentioned that these should be run weekly as a cron job (as is the case on main branch); now that they're working should I move these jobs to a separate file to run as a weekly cron?

**Comment by pllim** [comment]
> these jobs to a separate file to run as a weekly cron

I don't think so. As back in Travis CI days, we only run cron for the dev branch. CI for release branches are run as needed (during backports and release). So if we were to keep that workflow, all the jobs should be in one file.

**Comment by pllim** [comment]
Eh... Why did it auto-close?

**Comment by pllim** [comment]
Oh, did I click the wrong button? Ops... re-opened...

**Comment by embray** [comment]
Ok, thanks.  I was thinking the same, given that it's run infrequently anyways.

**Comment by embray** [comment]
Also how did the CI action magically start again?  I didn't push anything and it doesn't look like you did either.  Maybe because you closed/reopened?

**Comment by pllim** [comment]
Kudos on fixing the "exotic" archs. If applicable, maybe should forward-port those fixes if they fix one or more these: #11339 #11026 #10921 (might have missed some)

As for CI restarting, yes, re-open counts in their books.

**Comment by embray** [comment]
>  If applicable, maybe should forward-port those fixes if they fix one or more these: #11339 #11026 #10921 (might have missed some)

I did not note any of these failures. 

But actually I don't know why the configuration of this job should be any different on main than on this branch, or why it worked there but not here.  I agree it would be good to try to copy these changes back up, assuming they work.  I'll give it a try.

**Comment by pllim** [comment]
> Not sure why clocale had matplotlib pinned to 2.1.x

Looks like I had that in the very first commit of this PR and I have no idea why Past Self ™️ did that. I, too, cannot find this in the log of dev branch. Might have been a temporary hack to work around some failure at that time, but I cannot tell you why... Sorry. 😢

**Comment by pllim** [comment]
There is no good reason why the CI between dev and LTS diverged other than we changed things in dev and didn't backport it (either didn't get to it or deemed it not relevant for backport).

If you don't see the same failures between LTS and dev, then don't waste too much time going down that rabbit hole. The main goal of this PR is to port whatever we had for Travis CI at the time it stopped working to Actions. As long as that is happening, we can say the other stuff is out of scope. Hope this clarifies some things!

**Comment by pllim** [comment]
> Also why are you on GH on Sunday night? XD

Well, same question back at you. At least I wasn't pushing commits. 🤪

**Comment by embray** [comment]
There remains, at a minimum some failures with:

* 32-bit and parallel
* Python 3.6 with oldest supported version of all dependencies

It seems to be the same issue in both cases.

**Comment by embray** [comment]
Weird.  When I run that tox env locally I get failures too, but a completely different set of failures.

**Comment by embray** [comment]
I think it might be an issue of the branch needing to be rebased.

**Comment by pllim** [comment]
Yes, @astrofrog did a bunch of backports yesterday. Please rebase. Thanks!

**Comment by pllim** [comment]
Also, you might need to wait for the backport of #11391 to 4.0.x before you will pick it up here.

**Comment by embray** [comment]
> Also, you might need to wait for the backport of #11391 to 4.0.x before you will pick it up here.

I already added it.  Or does this happen automatically?  I'm still confused about how this is being done now.  I think it was better to have an explicit PR for backports.  That would prevent things like [this](https://github.com/astropy/astropy/pull/11123/commits/cda88309486568c216511ab012a84572c9ef922c) from happening.

**Comment by embray** [comment]
Ok, I see it's happened already, somehow or other.

**Comment by embray** [comment]
Now I am just getting some failures in astropy.modeling that seem to stem from the backport of #11177 as they are all related to tests on `Voigt1D`: https://github.com/astropy/astropy/runs/2141387509?check_suite_focus=true

**Comment by pllim** [comment]
Unless @dhomeier or @nden can suggest quick fix for the `modeling` failures today (Mar 18, 2021), I propose that we revert backport of #11177 .

As for the "xpass (strict)" failures in remote data job, we can just remove `xfail_strict = true` rule from `setup.cfg`, as I don't think anyone is going to care enough that some `xfail` test is now passing in LTS...?

**Comment by embray** [comment]
Unless someone explicitly needs it, I think it would be best to revert the backport of #11177 for now.  The differences between the expected and real results look pretty significant.  Might be hard to figure out what's going on.

**Comment by pllim** [comment]
Do you want to do that as part of this PR?

**Comment by embray** [comment]
Alternatively, we can just skip the tests of Voigt1D with quantities, since the other Voigt1D tests *are* passing :confused:

**Comment by dhomeier** [comment]
> Alternatively, we can just skip the tests of Voigt1D with quantities, since the other Voigt1D tests _are_ passing 😕

I'd certainly like to understand why parameter inputs with different, but equivalent units, do not work as supposed, and why that is specific to the new `Voigt1D` implementation (still, it might mean that the function is not working correcty in these cases).
```
actual = <Quantity [1.06217955, 1.06217955] Jy nm / Angstrom>
1128
desired = <Quantity [1.06217955, 1.06217955] Jy>, rtol = 1e-07, atol = None
```

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/11123#issuecomment-802134623 -- I'd rather we revert than xfail... FWIW

**Comment by embray** [comment]
And that's because of this: https://github.com/astropy/astropy/blob/3a3f0da1e12e0910b56c9b1a7b323e8ce322a0e7/astropy/units/quantity_helper/function_helpers.py#L50

**Comment by dhomeier** [comment]
Thanks @embray! I found that with current master and 4.0 with numpy 1.18 the same inputs return
` <Quantity 0.10621796 Jy nm / Angstrom>`
so the units are correctly factored in, but having them not reduced either is still ugly.
I was wondering if the `_parameter_units_for_data_units` somehow were not properly applied in the new version, perhaps they needed to be explicitly called when adding a custom `__init__`? But I could not find any examples for that...

**Comment by embray** [comment]
I think it should be a good enough fix to cast `w = w.to(u.dimensionless_unscaled)` there and leave it at that.  Testing now.

**Comment by dhomeier** [comment]
Did my last comment just disappear? Adding
```python
        if isinstance(z, Quantity):
            z = z.to(u.dimensionless_unscaled)
```
to `_wrap_wofz` let the tests pass as well, though when using different units for `fwhm_L` and `fwhm_G` still the ugly un-reduced units are returned. Would prefer to fix those as well, which probably would need to intercept the return value in `evaluate` and `fit_deriv`.

**Comment by embray** [comment]
@dhomeier Please review [this workaround](https://github.com/astropy/astropy/pull/11123/commits/7a79ed8afcde7922d36c8eb7af617323a1584dd6) if you get a chance.  I think it's good enough for now.

If you want, you may want to open a separate issue about the unit conversions in general to try to fix this in the main branch as well, though there the only problem is that the units are not being reduced, whereas here the problem is more severe.

**Comment by astrofrog** [comment]
@embray - just to reply to your comment about backports, we still do these manually and not through PRs (but I think there were some discussion about changing how we do backport in the infrastructure group)

File: astropy/modeling/functional_models.py
Comment (astrofrog) [inline astropy/modeling/functional_models.py:1200]: @embray - should this be done only if w is a Quantity? or is this guaranteed at this point?

Comment (embray) [reply astropy/modeling/functional_models.py]: No, It was a bug.  I just fixed it ^^

Comment (dhomeier) [reply astropy/modeling/functional_models.py]: No it's not - the functions accepts plain arrays as inputs everywhere. Checking for Quantity as in my belated comment is both necessary and sufficient.

**Comment by embray** [comment]
I'm seeing some :heavy_check_mark: here :grin:

**Comment by dhomeier** [comment]
Looks OK with the Quantity check; still pondering wether this is a better place for the test than `_wrap_wofz`...
Your solution would have the widest possible effect, wheres putting it outside would keep `_hum2zpf16c` closest to Franz Schreier's original implementation; also it is by definition a function of a dimensionless number.

**Comment by embray** [comment]
@dhomeier Please open a PR when you get a chance for a better fix.  I'm running out of time here that I can work on this.

**Comment by embray** [comment]
Anyone know anything about this SAMP failure?

```
astropy.samp.errors.SAMPProxyError: <SAMPProxyError 5: 'Private-key e08fe4be-8826-11eb-9f18-9bedc64c72c6 expired or invalid.'>
```

I don't think it has anything to do with the other one I fixed.

**Comment by pllim** [comment]
Ugh... It's always something with SAMP. I haven't seen that error but for the sake of the release, just `xfail` that one unless @astrofrog can come up with something.

**Comment by astrofrog** [comment]
I don't really have time to dig into the SAMP failure so either restart the CI to see if it's non-deterministic, or xfailing seems sensible.

**Comment by embray** [comment]
Yeah I was thinking of restarting it too.

**Comment by dhomeier** [comment]
> @dhomeier Please open a PR when you get a chance for a better fix. I'm running out of time here that I can work on this.

This looks good to me, not really sure actually which variant I would prefer.
Just one other suggestion to change the `FUNC_MODELS` value to
```diff
 {'class': Voigt1D,
  'parameters': {'amplitude_L': 2 * u.Jy, 'x_0': 505 * u.nm,
-                'fwhm_L': 100 * u.AA, 'fwhm_G': 50 * u.AA},
+                'fwhm_L': 100 * u.AA, 'fwhm_G': 5000 * u.pm},
  'evaluation': [(0.51 * u.micron, 1.0621795524 * u.Jy)],
  'bounding_box': False},
```
to check the numerical result if all inputs have different units.

**Comment by embray** [comment]
@pllim with the "(Allowed Failure) Python 3.7 with remote data and dev version of key dependencies" I know you mentioned something about xfail_strict = false, but since it's titled "Allowed Failure" isn't allowed to fail for now?

**Comment by pllim** [comment]
> since it's titled "Allowed Failure" isn't allowed to fail for now?

Sure... Sometimes ignoring it masks real bugs but if it is only SAMP, I guess we can ignore.

**Comment by embray** [comment]
Except the SAMP one wasn't on that :/

**Comment by embray** [comment]
Looks like the SAMP error was indeed a random fluke.

**Comment by eteq** [comment]
I'm confused by those IERS xpasses, because I'm not seeing that locally... can anyone else reproduce outside of CI?

**Comment by embray** [comment]
Sorry folks, I'm gonna have to bow out for tonight. We're still just waiting on the ARM test, but it was working last time I ran it, and it looks like the zArchitecture test passed. I'm not top worried about the allowed-fail test, though as @eteq says it is weird. If anyone wants to add a workaround to it they're welcome to. Otherwise I'm happy with where things stand, and can address the remaining issue in a follow up.

**Comment by pllim** [comment]
> I'm confused by those IERS xpasses

Turns out we simply skip them in dev now, so I manually backported those skips here.

I think that's the remaining issues, given that the rest of CI is now green? Can we merge when CI is green?

**Comment by pllim** [comment]
p.s. This is still technically my PR and I cannot approve it.

**Comment by pllim** [comment]
CI is green! Thanks, @embray ! @eteq or @astrofrog , please approve if you okay with this.

**Review by astrofrog — APPROVED** [review]
This is fantastic, thank you @pllim and @embray!

I notice a few manual backports - I think this is fine as-is but if the changelog consistency checks trip up because of these I will re-milestone those PRs and mark them as manual backports.

**Comment by pllim** [comment]
The two that I did in the last minute was:

* #10700 -- already in 4.2
* #10757 -- already in 4.2

Therefore, I believe they will not be listed by any backport scripts. If they are, then something is wrong.

**Pull request merged**
